x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,334 @@
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    public List<String> rulesetIds() {
+        return rulesetIds;
+    }
+
+    public Map<String, Object> matchCriteria() {
+        return matchCriteria;
+    }
+
+    public QueryBuilder organicQuery() {
+        return organicQuery;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(ORGANIC_QUERY_FIELD.getPreferredName(), organicQuery);
+        builder.startObject(MATCH_CRITERIA_FIELD.getPreferredName());
+        builder.mapContents(matchCriteria);
+        builder.endObject();
+        builder.startArray(RULESET_IDS_FIELD.getPreferredName());
+        for (String rulesetId : rulesetIds) {
+            builder.value(rulesetId);
+        }
+        builder.endArray();
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
+            throw new IllegalArgumentException("Can't have both pinned ids and pinned docs");
+        }
+
+        if (pinnedIds != null && pinnedIds.isEmpty() == false) {
+            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
+            return pinnedQueryBuilder.toQuery(context);
+        } else if (pinnedDocs != null && pinnedDocs.isEmpty() == false) {
+            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
+            return pinnedQueryBuilder.toQuery(context);
+        } else {
+            return organicQuery.toQuery(context);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        if (pinnedIds != null || pinnedDocs != null) {
+            return this;
+        } else if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
+            List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
+            List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
+            if (identifiedPinnedIds == null && identifiedPinnedDocs == null) {
+                return this; // not executed yet
+            } else {
+                return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetIds, identifiedPinnedIds, identifiedPinnedDocs, null, null);
+            }
+        }
+
+        // Identify matching rules and apply them as applicable
+        String rulesetId = rulesetIds.get(0);
+        GetRequest getRequest = new GetRequest(QueryRulesIndexService.QUERY_RULES_ALIAS_NAME, rulesetId);
```

Comment:
I changed this from a Query Rules API call to a Get request for a few reasons. One of which, was that the `RuleQueryBuilder` supports mocking Get requests, so it's a smaller change. 

This has a consequence though: Now, the yaml tests are failing with: 
```
"action [indices:data/read/get] is unauthorized for user [entsearch-user] with effective roles [user] on restricted indices [.query-rules]
```

I can change this back to a query rulesets API call, which will involve some updates to how we process query builder tests, or we can think about permissions here. @ioanatia and @jimczi curious to hear your thoughts, thanks

Reply:
The yaml test failure indicates that the credentials that are used when sending the search request need read access on the `.query-rules` index - this should not happen.
I think we should potentially try to avoid doing a GET document request directly on the system index here - but if we cannot avoid it, then we need to change it so that the GET document request uses the ENT_SEARCH origin.

We want to use something like `Client clientWithOrigin = new OriginSettingClient(client, ENT_SEARCH_ORIGIN);` and then instead of doing ` client.get(getRequest, ...` we would use `clientWithOrigin.get(getRequest, ...`.

Reply:
Thanks Ioana! I don't think we can/want to use the query rules API because it will require a dependency on the tests to broaden the allowed requests to something that's in an XPack module. I'm going to implement your suggestion to use the client with origin, I appreciate the feedback and suggestion!